src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -23,128 +23,300 @@
-
-    private final Runnable update;
-    private final Runnable reset;
+    private volatile ScheduledFuture<?> updateScoresScheduler;
+    private volatile ScheduledFuture<?> updateSamplesScheduler;
 
     public DynamicEndpointSnitch(IEndpointSnitch snitch)
     {
         this(snitch, null);
     }
 
-    public DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
+    protected DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
     {
         mbeanName = "org.apache.cassandra.db:type=DynamicEndpointSnitch";
         if (instance != null)
             mbeanName += ",instance=" + instance;
         subsnitch = snitch;
-        update = new Runnable()
+
+        if (DatabaseDescriptor.isDaemonInitialized())
         {
-            public void run()
-            {
-                updateScores();
-            }
-        };
-        reset = new Runnable()
+            applyConfigChanges(DatabaseDescriptor.getDynamicUpdateInterval(),
+                               DatabaseDescriptor.getDynamicSampleUpdateInterval(),
+                               DatabaseDescriptor.getDynamicBadnessThreshold());
+            open();
+        }
+    }
+
+    /**
+     * Allows subclasses to inject new ways of measuring latency back to this abstract base class.
+     */
+    protected interface ISnitchMeasurement
+    {
+        void sample(long value);
+        double measure();
+        Iterable<Double> measurements();
+    }
+
+    /**
+     * Adds some boookeeping that the DES uses over top of the various metrics techniques used by the
+     * implementations. This is used to allow CASSANDRA-14459 latency probes as well as further experimentation
+     * on new latency measurement techniques in CASSANDRA-14817
+     *
+     * {@link AnnotatedMeasurement#millisSinceLastRequest} is set to zero through
+     * {@link DynamicEndpointSnitch#sortedByProximity(InetAddressAndPort, ReplicaCollection)}. It defaults to
+     * the maximum interval so that we only start probing once it has been ranked at least once
+     * {@link AnnotatedMeasurement#millisSinceLastMeasure} is set to zero from
+     * {@link DynamicEndpointSnitch#receiveTiming(InetAddressAndPort, long, LatencyMeasurementType)}.
+     *
+     * {@link AnnotatedMeasurement#millisSinceLastMeasure and {@link AnnotatedMeasurement#probeTimerMillis}
+     * are incremented via {@link DynamicEndpointSnitch#updateSamples()}
+     */
+    protected static class AnnotatedMeasurement
+    {
+        // Used to optimally target latency probes only on nodes that are both requested for ranking
+        // and are not being measured. For example with token aware clients a large portion of the cluster will never
+        // be ranked at all and therefore we won't probe them.
+        public AtomicLong millisSinceLastRequest = new AtomicLong(MAX_PROBE_INTERVAL_MS);
+        public AtomicLong millisSinceLastMeasure = new AtomicLong(0);
+        public long probeTimerMillis = 0;
```

Comment:
This would make a lot more sense to me as nextProbeDelay? Since we are tracking how far into the future to delay the next probe?

Reply:
Done.